iceberg: action for removing snapshots from table metadata

andrwng opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->
This introduces snapshot removal as an action that can be performed on table metadata. This action will compute snapshots and snapshot references to remove from the table metadata. The rules for retention are described in `remove_snapshots_action.h`, and follow the upstream implementation[1].

Note that it will not perform IO to remove the relevant files in S3. This PR also does not call this new code from the coordinator. Those will come in later PRs.

[1] https://github.com/apache/iceberg/blob/72dcce95e294835f978dc1d6c9a3be5d89123410/core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

andrwng committed:
iceberg: add remove-snapshot-ref update

This table update is going to be plugged into an upcoming transaction to
expire snapshots. As a part of expiring snapshots, we also need to clean
up snapshot references that become defunct.

andrwng committed:
iceberg: action to remove snapshots

This adds a building block towards lowering the size of table metadata.
Table metadata currently keeps all snapshots without ever removing them.
This means that table metadata will grow unbounded if left as is.

Previous commits already added the appropriate updates and requirements
for snapshot and snapshot reference removal. This commit just adds an
action to actually compute what snapshots to remove.

The logic follows that implemented by the Apache Iceberg RemoveSnapshots
action.

Note, removal in this commit doesn't delete the underlying files in the
bucket, but rather tells the catalog to remove the snapshots from
metadata. A separate task needs to clean up the removed, orphaned
snapshots.

andrwng committed:
iceberg: transactional interface for snapshot removal

Adds the new snapshot removal action as a method of the Iceberg
transactions interface.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/60579-->
<!--9d81ab1c-01ea-4695-b9e3-cacc3ea10e79-->
<!--title-->
## Retry command for [Build#60579](https://buildkite.com/redpanda/redpanda/builds/60579)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":true,"mixed_versions":true,"with_iceberg":false,"with_tiered_storage":true}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--ace8bba8-7bcf-45b3-8e2f-5f80123551db-->
<!--TEST RESULTS-->
<!--538567c3-2fe8-4a2f-9c77-cf20de9e37d0-->
<!--TEST RESULTS-->
<!--f27e644b-0ffe-4dcc-9040-b5c173b48a3d-->
<!--TEST RESULTS-->
<!--c204364d-af6b-4505-9c77-5224c8175d88-->
<!--TEST RESULTS-->
<!--a66138ef-74f3-41a1-9bf3-0227323da16d-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60579">build#60579</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rm_stm_tests_rpunit.rm_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60579#01944eea-667a-47f2-a3c3-819ad409d247 | FLAKY | 1/2 |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli | ducktape | https://buildkite.com/redpanda/redpanda/builds/60579#01944f30-d307-4452-99ea-ee0df6253912 | FLAKY | 2/6 |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli | ducktape | https://buildkite.com/redpanda/redpanda/builds/60579#01944f43-7ae5-41cc-80dc-1d30484bee6a | FLAKY | 4/6 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=True.with_tiered_storage=True.with_iceberg=False.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/60579#01944f30-d308-4044-9ae7-c3aeaca5b315 | FAIL | 0/1 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60602">build#60602</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli | ducktape | https://buildkite.com/redpanda/redpanda/builds/60602#01945159-633a-41ec-a8fc-7a8e6580f25a | FLAKY | 3/6 |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli | ducktape | https://buildkite.com/redpanda/redpanda/builds/60602#0194515e-1f51-42d0-b4cc-35d80ed5fbd3 | FLAKY | 4/6 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60758">build#60758</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| idempotency_tests_rpunit.idempotency_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60758#0194690b-2dfd-4d61-8e15-e488e741e945 | FLAKY | 1/2 |
| idempotency_tests_rpunit.idempotency_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60758#0194690b-2dfe-4320-bf68-a4f8569173e7 | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/60758#0194694c-1bce-4973-a958-e308cab194e8 | FLAKY | 5/6 |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli | ducktape | https://buildkite.com/redpanda/redpanda/builds/60758#0194694c-1bcf-47b3-92b4-552bc22c498b | FLAKY | 3/6 |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli | ducktape | https://buildkite.com/redpanda/redpanda/builds/60758#01946951-34a4-4982-8cc8-3e26f2fea2be | FLAKY | 3/6 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60800">build#60800</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| coordinator_rpunit.coordinator_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60800#01946b4a-71ef-4603-97dd-08b89463db8b | FLAKY | 1/2 |
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60800#01946b4a-71f0-4a19-a454-c819347c5aa6 | FLAKY | 1/2 |
| partition_balancer_planner_test_rpunit.partition_balancer_planner_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60800#01946b4a-71ef-4603-97dd-08b89463db8b | FLAKY | 1/2 |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli | ducktape | https://buildkite.com/redpanda/redpanda/builds/60800#01946b8c-c287-49af-a8cd-b05b64d99f50 | FAIL | 0/1 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60870">build#60870</a></summary>


| test_id                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   |
|:--------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| partition_balancer_simulator_test_rpunit.partition_balancer_simulator_test_rpunit                 | unit        | https://buildkite.com/redpanda/redpanda/builds/60870#019470e4-47ab-4f80-9d48-85cc4f7aa2b7 | FLAKY         | 1/2      |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60870#01947126-b4e8-4366-b672-1e258b091a96 | FLAKY         | 1/2      |
</details>
<!--/body-->

andrwng committed:
iceberg: no-op on commit when no updates

Updates the catalog implementations to return early when a given
transaction is empty. This will be useful in the near future since
snapshot removal may repeatedly result in no new snapshots to remove,
and committing such transactions isn't inherently problematic.

ztlpn commented on src/v/iceberg/transaction.h:
```diff
@@ -40,6 +40,10 @@ class transaction {
 
+    // Removes expired snapshots from the table. Note, this does not perform IO
+    // to delete any now-orphaned metadata or data.
+    ss::future<txn_outcome> remove_expired_snapshots(model::timestamp);
```
nit: adding a descriptive parameter name would help with understanding what timestamp this is (i.e. "now"?).

andrwng replied:
Done

ztlpn commented on src/v/iceberg/remove_snapshots_action.cc:
```diff
@@ -0,0 +1,305 @@
+    try {
+        return std::stol(*p_str);
+    } catch (...) {
+        vlog(log.warn, "Invalid long for property '{}': '{}'", prop, *p_str);
```
nit: makes sense to print table location as well

andrwng replied:
Done

ztlpn commented on src/v/iceberg/remove_snapshots_action.cc:
```diff
@@ -0,0 +1,305 @@
+    return get_long_property(table, max_ref_age_ms_prop)
+      .value_or(remove_snapshots_action::default_max_ref_age_ms);
+}
+size_t table_min_snapshots_retained(const table_metadata& table) {
```
nit: I wonder if we should use `int64_t` here, as size_t in theory could be something else... Though I couldn't find the mandated int type for this property (the spec says per-ref `min-snapshots-to-keep` is `int` but says nothing about the table property)

andrwng replied:
Hum, good catch. The Java implementation actually represents this as an int -- will update

andrwng replied:
Done

ztlpn commented on src/v/iceberg/remove_snapshots_action.cc:
```diff
@@ -0,0 +1,305 @@
+    auto next_snap = std::make_optional<snapshot_id>(snap_id);
+    while (next_snap.has_value()) {
+        auto snap_it = std::ranges::find(
+          table.snapshots.value(), next_snap.value(), &iceberg::snapshot::id);
```
I wonder if we should have by-id maps for objects in table_metadata, this can easily become accidentally quadratic.

andrwng replied:
Yeah that's fair. I'll add a method to the table metadata to build a map. Maybe longer term we'd want some kind of lazy data member, but for the immediate usage, just translating to a map is more than enough

ztlpn commented on src/v/iceberg/remove_snapshots_action.cc:
```diff
@@ -0,0 +1,305 @@
+            referenced_snaps.emplace(ref.snapshot_id);
+        }
+    }
+    vlog(log.debug, "{} snapshots are referenced", referenced_snaps.size());
```
nit: table location for context

andrwng replied:
Done

ztlpn commented on src/v/iceberg/remove_snapshots_action.h:
```diff
@@ -0,0 +1,76 @@
+// which is configurable per table.
+//
+// NOTE: this does not perform IO to purge the corresponding manifest lists or
+// the underlying manifests and data files. Callers are expected to do so upon
```
BTW do you know if `snapshot-log` and `metadata-log` arrays in the table metadata are cleaned automatically and who should clean them if not?

andrwng replied:
The fields are meant to be updated every time a transaction is performed -- i.e. we would apply the action and then recompute `snapshot-log` and `metadata-log` (and we'd then purge metadata per the new metadata log).

For snapshot log: https://github.com/apache/iceberg/blob/d96901b843395fe669f6bd4f618f8e5e46c0eed4/core/src/main/java/org/apache/iceberg/TableMetadata.java#L1509-L1510
For metadata log: https://github.com/apache/iceberg/blob/d96901b843395fe669f6bd4f618f8e5e46c0eed4/core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java#L126

For the object storage catalog, it would definitely be on us to compute the right logs and perform the appropriate cleanup (once we implement the fields, of course). What I'm unsure of is whether the catalog would perform these actions for us

ztlpn commented on src/v/iceberg/transaction.h:
```diff
@@ -40,6 +40,10 @@ class transaction {
 
+    // Removes expired snapshots from the table. Note, this does not perform IO
+    // to delete any now-orphaned metadata or data.
+    ss::future<txn_outcome> remove_expired_snapshots(model::timestamp);
```
So I peeked into the next PR and it seems that to determine the list of expired snapshots we have to do a rather unpleasant load-before, load-after sequence. I wonder if the transaction could return the set of expired snapshots directly.

andrwng replied:
It could, I had that implemented initially too, but I ended up going with this because this is what the Java implementation does. The theoretical concern is that the catalog may return a different set of snapshots to remove, or may end up removing more than just what this transaction computed. It seems debatable whether Redpanda should honor its own removals vs those given by the catalog, and I opted to give more weight to the catalog. Happy to discuss further if you feel strongly about this.

ztlpn replied:
No, not feeling particularly strongly, it makes sense to follow what the reference impl does.

andrwng committed:
iceberg: add method for building id->snapshot map

Callers that need to do non-trivial computation on the table metadata
shouldn't be tempted to use the existing vector in the metadata (left
simple to keep serialization simple).

This commit adds a method to build a map from the snapshots vector keyed
by the snapshot id to make it less tempting to write quadratic code.

andrwng committed:
iceberg: use table properties to dictate snapshot removal

Adds parsing of the Iceberg table properties that are relevant to
snapshot removal. This will make it less surprising if users have these
properties set, and will also make it easier to write small-scale tests
(the default expiry is 5 days).

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
Failed to create a backport PR to v24.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-24767-v24.3.x-816 remotes/upstream/v24.3.x
git cherry-pick -x dde376054d 3cb7e77fbf 54a9adf63e dc365b0fd7 b588d8569d ce03868de2
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/12819965530).
